src/Datastore/RAMDataStore.cpp

```diff
@@ -0,0 +1,337 @@
+  };
+  // actually remove all expired sessions
+  mUsers.erase(remove_if(mUsers.begin(), mUsers.end(), checkExpired),
+               mUsers.end());
+
+  return nullopt;
+}
+
+TResultOpt RAMDataStore::addTrack(BaseTrack track, QueueType q) {
+  // Exclusive Access to Song Queue
+  unique_lock<shared_mutex> MyLock(mQueueMutex, defer_lock);
+  MyLock.lock();
+
+  // select Queue
+  Queue *pQueue = SelectQueue(q);
+  if (pQueue == nullptr) {
+    return Error(ErrorCode::InvalidValue, "Invalid Parameter in Queue");
+  }
+
+  // check for existing Track
+  QueuedTrack qtr;
+  qtr.trackId = track.trackId;
+  auto it = find(pQueue->tracks.begin(), pQueue->tracks.end(), qtr);
+  if (it == pQueue->tracks.end()) {
+    // Track is unique, insert it into vector
+    qtr.title = track.title;
+    qtr.album = track.album;
+    qtr.artist = track.artist;
+    qtr.duration = track.duration;
+    qtr.iconUri = track.iconUri;
+    qtr.addedBy = track.addedBy;
+    qtr.votes = 0;
+    qtr.insertedAt = time(nullptr);
+    qtr.LastPlayedxSongsAgo =
+        1;  // set to one so that votes of songs count immediately after
+            // insertion. See nextTrack and VoteTrack and Tracks operator<
+    pQueue->tracks.push_back(qtr);
+    return nullopt;
+  } else {
+    return Error(ErrorCode::AlreadyExists, "Track already exists");
+  }
+}
+
+TResult<BaseTrack> RAMDataStore::removeTrack(TTrackID ID, QueueType q) {
+  // Exclusive Access to Song Queue
+  unique_lock<shared_mutex> MyLock(mQueueMutex, defer_lock);
+  MyLock.lock();
+
+  // select Queue
+  Queue *pQueue = SelectQueue(q);
+  if (pQueue == nullptr) {
+    return Error(ErrorCode::InvalidValue, "Invalid Parameter in Queue");
+  }
+
+  // delete Track
+  QueuedTrack track;
+  track.trackId = ID;
+  auto it = find(pQueue->tracks.begin(), pQueue->tracks.end(), track);
+  if (it == pQueue->tracks.end()) {
+    return Error(ErrorCode::DoesntExist, "Track doesnt exist in this Queue");
+  } else {
+    track = *it;
+    // Track is there, delete it from vector
+    pQueue->tracks.erase(it, it);
+    return track;
+  }
+}
+
+TResult<bool> RAMDataStore::hasTrack(TTrackID ID, QueueType q) {
+  // Shared Access to Song Queue
+  shared_lock<shared_mutex> MyLock(mQueueMutex, defer_lock);
+  MyLock.lock();
+
+  // select Queue
+  Queue *pQueue = SelectQueue(q);
+  if (pQueue == nullptr) {
+    return Error(ErrorCode::InvalidValue, "Invalid Parameter in Queue");
+  }
+
+  // find Track
+  QueuedTrack track;
+  track.trackId = ID;
+  auto it = find(pQueue->tracks.begin(), pQueue->tracks.end(), track);
+  if (it == pQueue->tracks.end()) {
+    return false;
+  } else {
+    return true;
+  }
+}
+
+TResultOpt RAMDataStore::voteTrack(TSessionID sID, TTrackID tID, TVote vote) {
+  // Exclusive Access to Song Queue and User
+  unique_lock<shared_mutex> MyLockQueue(mQueueMutex, defer_lock);
+  unique_lock<shared_mutex> MyLockUser(mUserMutex, defer_lock);
+  scoped_lock(MyLockQueue, MyLockUser);
```

Comment:
`scoped_lock` creates an object which locks the mutexes until it goes out of scope, which is instantly since you do not store the object in a variable

Reply:
you probably want `std::lock`. in this case `defer_lock` is fine
https://en.cppreference.com/w/cpp/thread/lock

Reply:
thx, fixed